refactor(postgresql.conf): configure postgresql.conf to use 'include_dir' directive

hunleyd opened the pull request:
> What kind of change does this PR introduce?

Moved all config files that are included in the postgresql.conf into the /etc/postgresql-custom dir then updated postgresql.conf to use include_dir instead of include. To ensure the same ordering of the various GUCs (to ensure proper overriding of values), rename files with digits prepended.
Finally, split auto_explain and pg_cron out into their own config files instead of them being tacked on the end of the main PG config.

this is take 2, which should fix the supautils [issue](https://supabase.slack.com/archives/C02FHG9QQAF/p1760112949721399)

hunleyd committed:
refactor(postgresql.conf): configure postgresql.conf to use 'include_dir' directive and move logging.conf into the postgresql-customer dir

hunleyd committed:
refactor(ansible): rename all conf files in /etc/postgresql-custom to preserve ordering

hunleyd committed:
refactor(postgresql): split auto_explain and pg_cron out into their own config files

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres: (29 commits)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1818)
  fix: update Dockerfiles for changes to postgis multiversion (#1817)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1813)
  feat: support multiple versions of the postgis extension (#1667)
  Update pgsql-http to handle semver (#1816)
  feat: add retry policy for auth service routes (#1782)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1811)
  fix: move tmpdir for SAA to one that always exists (#1799)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1810)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1809)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1807)
  build: automatically cancel old tests/build on new push (#1808)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1804)
  docs: small adjust (#1806)
  docs: getting started guide in README (#1805)
  fix: templating of SAA service file
  fix: adjust the qemu vars filename
  refactor(ansible): bring our ansible up to modern ansible-lint standards
  feat: support multiple versions of the http extension (#1664)
  feat: supply a slightly different supabase-admin-agent configuration for qemu artifacts
  ...

hunleyd committed:
fix(nix): it helps to use the correct filename everywhere

hunleyd committed:
refactor(postgresql.conf): use relative path to include_dir

hunleyd committed:
fix(postgresql.conf): revert the location of include_dir back to absolute path

hunleyd committed:
refactor(postgresql): switch to 'include_dir' and then rename conf files to ensure ordering

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  fix(nix): remove '%' character from Nix trusted-public-keys configuration (#1840)
  chore: systemd hardening (#1837)
  Revert "refactor(postgresq): switch to 'include_dir' and then rename conf files to ensure ordering" (#1835)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1821)
  fix(pgmq): replace drop_queue function if exists (#1828)
  refactor(postgresq): switch to 'include_dir' and then renames conf files to ensure ordering (#1820)
  feat: support multiple versions of the pg_jsonschema extension (#1757)
  feat: bump auth to v2.180.0 (#1829)
  feat: update supautils confs w/ new tables
  feat(migrations): predefined role grants (#1815)
  feat: run pg_regress tests after installing the last version of the extension (#1826)
  fix(pgmq): add missing helper function in migration script (#1825)
  feat: support multiple versions of the pgmq extension (#1668)
  feat: run pg_regress during extension tests (#1812)
  feat: support multiple versions of the pgrouting extension (#1687)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1819)
  feat: support multiple versions of the pg-graphql extension (#1761)

hunleyd committed:
refactor(docker): don't sed, just echo

samrose committed:
fix: test harness needed to handle config files correctly

hunleyd committed:
refactor(PG-configuration): move includes to conf.d and rename wth the numeric prefixes as per Sam

hunleyd committed:
fix(nix): nix doesn't like it when you try to assign vars to undef'd vars

hunleyd committed:
fix: update path to 05-supautils

hunleyd committed:
fix: update path to 00-logging

hunleyd committed:
refactor(postgresql): mv 00-logging.conf to just logging.conf

hunleyd committed:
refactor(postgres): 01-generated-optimizations -> generated-optimizations

hunleyd committed:
refactor(postgresql): 02-custom-overrides -> custom-overrides

hunleyd committed:
refactor(postgres): 03-wal-g -> wal-g

hunleyd committed:
refactor(postgres): 04-read-replica -> read-replica

hunleyd committed:
refactor(postgres): 05-supautils -> supautils

hunleyd committed:
style(postgresql): style fixups

hunleyd committed:
fix: more path updates

samrose committed:
Indata 152 run server (#1850)

* fix: refactor tests to work with include_dir approach
and be more accurate with production conditions

* tests: use distinct savepoint names in test

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1852)
  chore: bump to release (#1855)
  docs: getting started nix install or config revised (#1853)
  docs: how to update new structure (#1851)
  feat: support multiple versions of the plpgsql_check extension (#1684)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1848)
  feat(wrappers): support more versions (#1831)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1833)
  fix: fine tune protection rules to unblock wal-g functionality (#1846)
  fix: incorporate v3.0.0 supautils (#1844)
  fix: remove git revision from postgres package

samrose commented:
@hunleyd this is still in draft mode of course. But requesting that once we are ready for review, we need @soedirgo @steve-chavez @za-arthur and myself to review before we merge and deploy just due to all of the changes.

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1861)
  fix: bump admin-api with enforce ssl fix (#1857)
  feat: support multiple versions of the pg_repack extension (#1688)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1856)

hunleyd commented:
looks like it fails on:
```
2025-10-24T18:05:08.0704634Z error: Cannot build '/nix/store/6ppacf0lg2xjbgw5f08bv3z0z95y1l0g-postgres-17_11-check-harness.drv'.
2025-10-24T18:05:08.0708489Z        Reason: builder failed with exit code 1.
2025-10-24T18:05:08.0709732Z        Output paths:
2025-10-24T18:05:08.0711914Z          /nix/store/i2cw7k1wnlfffmd68gyn5d26b4fvy1j1-postgres-17_11-check-harness
2025-10-24T18:05:08.0713654Z        Last 25 log lines:
2025-10-24T18:05:08.0714742Z        >
2025-10-24T18:05:08.0715708Z        > pg_regress tests failed
2025-10-24T18:05:08.0719432Z        > diff -U3 /nix/store/a58i3yh31cjad0lgmyx4p7xy9zmg3l13-tests/expected/postgres_fdw.out /nix/store/i2cw7k1wnlfffmd68gyn5d26b4fvy1j1-postgres-17_11-check-harness/regression_output/results/postgres_fdw.out
2025-10-24T18:05:08.0724613Z        > --- /nix/store/a58i3yh31cjad0lgmyx4p7xy9zmg3l13-tests/expected/postgres_fdw.out       1970-01-01 00:00:01.000000000 +0000
2025-10-24T18:05:08.0729147Z        > +++ /nix/store/i2cw7k1wnlfffmd68gyn5d26b4fvy1j1-postgres-17_11-check-harness/regression_output/results/postgres_fdw.out     2025-10-24 18:05:05.290257489 +0000
2025-10-24T18:05:08.0732085Z        > @@ -14,10 +14,9 @@
2025-10-24T18:05:08.0733091Z        >  set role postgres;
2025-10-24T18:05:08.0734286Z        >  -- postgres_fdw should be owned by the superuser
2025-10-24T18:05:08.0736742Z        >  select fdwowner::regrole from pg_foreign_data_wrapper where fdwname = 'postgres_fdw';
2025-10-24T18:05:08.0738481Z        > -    fdwowner
2025-10-24T18:05:08.0739419Z        > -----------------
2025-10-24T18:05:08.0740424Z        > - supabase_admin
2025-10-24T18:05:08.0741376Z        > -(1 row)
2025-10-24T18:05:08.0742248Z        > + fdwowner
2025-10-24T18:05:08.0743121Z        > +----------
2025-10-24T18:05:08.0743821Z        > +(0 rows)
2025-10-24T18:05:08.0744133Z        >
2025-10-24T18:05:08.0744963Z        >  -- Verify that `postgres` can use the FDW despite not owning it
2025-10-24T18:05:08.0745499Z        >  create server s
2025-10-24T18:05:08.0745837Z        > @@ -27,4 +26,5 @@
2025-10-24T18:05:08.0746176Z        >      port '5432',
2025-10-24T18:05:08.0746547Z        >      dbname 'postgres'
2025-10-24T18:05:08.0746898Z        >    );
2025-10-24T18:05:08.0747307Z        > +ERROR:  foreign-data wrapper "postgres_fdw" does not exist
2025-10-24T18:05:08.0747804Z        >  rollback;
2025-10-24T18:05:08.0748127Z        For full logs, run:
2025-10-24T18:05:08.0748724Z          nix log /nix/store/6ppacf0lg2xjbgw5f08bv3z0z95y1l0g-postgres-17_11-check-harness.drv
```

is that what you meant by 'tests still need some work' @samrose ?

samrose commented:
@hunleyd no the tests were expected to pass. But probably whatever is causing the issue is simple. I'll take a look and report back

samrose committed:
fix: fix up postgres_fdw test

hunleyd committed:
fix(setup-postgres): it helps to remove *all* conflict markers before pushing

hunleyd committed:
Merge branch 'INDATA-152' of github.com:supabase/postgres into INDATA-152

* 'INDATA-152' of github.com:supabase/postgres:
  fix: fix up postgres_fdw test

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  chore: add our substituter config to flake.nix (#1839)
  refactor: improve test harness logging and error handling (#1834)
  feat: support multiple versions of the pgroonga extension (#1677)
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1862)
  fix: bump version for new release (#1860)

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres:
  feat(Nix-flakes): Add pgBackRest flake (#1859)
  fix: wrappers 0.5.3 missing (#1872)

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres:
  fix: covering migrations for wrappers across all versions` (#1876)

hunleyd committed:
test(pgroonga): update the expected test output for pgroonga

hunleyd committed:
fix(setup-wal-g): remove errant stale task from a bad upstream merge

hunleyd committed:
fix(setup-wal-g): s/postgresq/postgres/

hunleyd committed:
test(ami-versions): bump the ami versions so i can run the infra tests

hunleyd committed:
test(ansible): revert my custom build numbers since the GHA all pass somehow ;)

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  feat: support multiple versions of the plv8 extension (#1676)
  chore: bump wrappers version 0.5.6 (#1877)
  feat: support multiple versions of the pg_hashids extension (#1755)

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  feat: update supautils (#1879)

hunleyd committed:
fix(prime.sql): typo

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  refactor(ansible): bring our ansible up to modern ansible-lint standards (#1869)

hunleyd committed:
test(pg_regress): add updated expected test outputs

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  feat: enable GitHub merge queues (#1849)

hunleyd committed:
Merge branch 'develop' into INDATA-152

hunleyd committed:
test(pg_regress): update expected output for pg15 plv8

hunleyd committed:
Merge branch 'INDATA-152' of github.com:supabase/postgres into INDATA-152

* 'INDATA-152' of github.com:supabase/postgres:
  fix: change trigger to `merge_group` in "Release Migrations - Staging / build (push)" workflow (#1884)
  tests: newline readme (#1883)

samrose committed:
Revert "test(pg_regress): update expected output for pg15 plv8"

This reverts commit bd512ff8b977b9c679c745d38efff817e883197f.

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  fix: we won't deploy this newest version of plv8 yet (#1885)

hunleyd committed:
Merge branch 'INDATA-152' of github.com:supabase/postgres into INDATA-152

* 'INDATA-152' of github.com:supabase/postgres:
  Revert "test(pg_regress): update expected output for pg15 plv8"

hunleyd committed:
fix(stage2-setup-postgres): supautils.conf is in conf.d now

samrose reviewed (changes requested):
Going to run through a multi step testing checklist with @jchancojr prior to merging this

Basically 

specific pg_regress tests for configuration

local infra testing of pause/restore, upgrade to this image, overall functionality

Run through internal  integration  of image 

A few review items from Sam on this PR that will be added to the actual PR here

jchancojr committed:
refactor(supautils.conf): fix conflicts

jchancojr committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

samrose committed:
chore: bump version

hunleyd committed:
Merge branch 'develop' into INDATA-152

jchancojr committed:
test(pg_regress): update pg_regress pg roles tests

jchancojr committed:
Merge branch 'develop' into INDATA-152

jchancojr commented:
getting supautils related errors 

```
postgres@db-wfcgdcogpuhyxlzxaiur:~$ psql
WARNING:  invalid configuration parameter name "supautils.disable_program", removing it
DETAIL:  "supautils" is now a reserved prefix.
```

jchancojr committed:
Merge branch 'develop' into INDATA-152

jchancojr commented:
`/data/pgdata/postgresql.conf` still exists. Should we symlink this to `postgresql-custom` or something?  @hunleyd

hunleyd commented:
> `/data/pgdata/postgresql.conf` still exists. Should we symlink this to `postgresql-custom` or something? @hunleyd

imho, we should, yes. or remove them entirely. all the `.conf` files in this dir (except `postgresql.auto.conf` obviously).

interested in @samrose 's thoughts though

hunleyd committed:
Merge branch 'develop' into INDATA-152

jchancojr committed:
Merge branch 'develop' into INDATA-152

hunleyd committed:
test(nix): update epected test outputs

samrose commented:
@hunleyd see if you agree with all of this

Recommending remove the symlink entirely and maintain your current approach of configs in /etc/postgresql/ with lots of testing to make sure nothing breaks:

```
  Why This is Best:

  1. PostgreSQL Official Documentation endorses separating configs from data
  2. Debian/Ubuntu Standard Practice (which your Ansible builds follow)
  3. Your current postgresql.conf.j2 already implements this correctly:
  data_directory = '/var/lib/postgresql/data'
  hba_file = '/etc/postgresql/pg_hba.conf'
  include_dir = '/etc/postgresql-custom/conf.d'
  4. The symlink serves no purpose - PostgreSQL is already configured to read from /etc/
```


  - PostgreSQL Official Docs: https://www.postgresql.org/docs/current/runtime-config-file-locations.html
  - include_dir Best Practices: https://www.postgresql.org/docs/current/config-setting.html
  
  
  Just please test as much as you can in local infra, and we should run e2e tests against your branch. Please create a playbook to document running e2e tests, but also reach out for help if you get stuck trying to do it. Thanks for the careful work on this.

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres:
  fix: don't run a nix installer on qemu image build (#1916)
  fix: restore ami release to us east 1 (#1913)
  feat: upgrade admin-mgr to 0.32.3 (#1915)
  fix: Prepare job (lines 23-25): Moved checkout before nix-install-ephemeral (#1914)
  feat: support multiple versions of the wal2json extension (#1875)

hunleyd committed:
fix: Remove extraneous postgresql config files

This change adds a task to remove `pg_hba.conf`, `pg_ident.conf`, and `postgresql.conf` from the PostgreSQL data directory.

This ensures there is no confusion about what configuration file is active.

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  Revert "feat: update supautils (#1879)" (#1912)

hunleyd committed:
feat: Symlink PostgreSQL config files

Symlinks PostgreSQL configuration files (`pg_hba.conf`, `pg_ident.conf`, `postgresql.conf`) from `/etc/postgresql/` into the PGDATA directory (`/var/lib/postgresql/data/`).

This ensures that the configuration files within PGDATA are always managed centrally and point to the correct, actively used configuration. The `force: true` option is added to overwrite any existing files, guaranteeing the symlink is correctly established.

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  feat: support multiple versions of the pg_plan_filter extension (#1873)

hunleyd committed:
fix: Fix PostgreSQL initdb and config symlinks

- Explicitly set the data directory using `--pgdata` in `pg_ctl initdb` commands.

hunleyd committed:
Merge branch 'develop' into INDATA-152

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres:
  fix: add alpn_protocols to envoy (#1503)

hunleyd committed:
fix(setup-postgres): switch from `cmd` to `argv` format to see if that gets rid of the arg errors

hunleyd committed:
Merge branch 'INDATA-152' of github.com:supabase/postgres into INDATA-152

* 'INDATA-152' of github.com:supabase/postgres:

hunleyd committed:
fix(setup-postgres): use the short form for the data dir option

hunleyd committed:
fix(setup-postgres): use the short form for the data dir option

hunleyd committed:
refactor: Update Postgres Ansible tasks and data dir
- Standardize Ansible module usage for PostgreSQL configuration files and related tasks.
- Update the default PostgreSQL data directory for `initdb` to `/data/pgdata`.

hunleyd committed:
chore: Update PostgreSQL initdb for Nix

This commit updates the Ansible tasks for setting up PostgreSQL.

The primary changes involve:
- Shifting the `initdb` executable path from `/usr/lib/postgresql/bin/pg_ctl` to `/var/lib/postgresql/.nix-profile/bin/initdb`, reflecting a Nix-based installation.
- Adjusting the arguments passed to `initdb` for clarity and correctness.
- Updating the `creates` path in the Ansible `args` to match the new data directory structure.

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres:
  feat: support multiple versions of the pg-safeupdate extension (#1871)
  feat(setup-postgres.yml): enable checksums via initdb for postgres (#1891)

hunleyd committed:
fix(setup-docker): copy supautils.conf in nixpkg_mode

hunleyd committed:
refactor: Improve PostgreSQL setup and initialization

Consolidate template tasks for postgresql.conf, pg_hba.conf, and pg_ident.conf into a single loop.

Ensure placeholder config files are correctly placed within `/etc/postgresql-custom/conf.d/`.

Remove the redundant `-D /data/pgdata` argument from the `initdb` commands, as the data directory is now managed by a symlink to `/data/pgdata`.

Add debug commands to verify the data directory symlink.

hunleyd committed:
Merge branch 'develop' of github.com:supabase/postgres into INDATA-152

* 'develop' of github.com:supabase/postgres:
  feat: support multiple versions of the pg_stat_monitor extension (#1683)
  feat: support multiple versions of the pgjwt extension (#1751)

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  chore: bump Wrappers version to 0.5.7 (#1945)
  chore: bump versions (#1949)
  Revert "docs: create an automated changlog file that shows up in release file…" (#1948)
  chore: remove precreated Storage objects (#1911)
  refactor: Move PostgreSQL auto_explain config to conf.d (#1943)
  docs: create an automated changlog file that shows up in release files (#1793)
  fix: removing dead nix code (#1946)
  docs: expansion to help understand anatomy and code api in user here (#1944)
  fix: try to use mirror fallbacks for docker images (#1942)
  feat: need this workflow present to test (#1941)
  feat: Add PostgreSQL conf.d directory (#1929)
  [jj-spr] initial version (#1866)
  chore: upgrade supabase-admin-agent (#1935)
  fix: garbage collect after every profile intall + remove osquery (#1938)
  fix(ci): limit max-jobs of nix to 8 to prevent OOM while running nix flake check (#1933)
  Revert "fix: disable temporary Nix checks on aarch64-linux" (#1934)
  fix: adaptive retry on aws (#1917)
  fix: disable temporary Nix checks on aarch64-linux (#1930)
  feat: release orioledb with rewind functionality (#1880)
  feat: reduciton of closure size for plv8 (#1925)

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  refactor: Remove wal-g v3 and simplify build (#1937)
  refactor: Refactor PostgreSQL cron config and deprecations (#1951)

hunleyd committed:
Merge remote-tracking branch 'origin/develop' into INDATA-152

* origin/develop:
  chore: prevent loopback ssh connections (#1959)
  chore: rename infra repo (#1958)
  chore: [ansible/vars] bumping admin-api version (#1960)
  fix: reserve the port for postgres_exporter with sysctl (#1953)
  chore: Update PostgREST setup and versions (#1955)
  Bump PostgREST to 14.1 (#1909)

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  feat: Ensure postgres conf.d directory is set up (#1961)
  Multi version ext/pg partman (#1854)

hunleyd committed:
Merge remote-tracking branch 'origin' into INDATA-152

* origin:
  refactor: Move read-replica.conf to conf.d (#1956)
  ci: Custom GitHub Runners for Nix Builds (#1745)
  fix: search path and migration grants (#1939)
  chore: bump admin api version (#1964)
